src/library/scala/collection/mutable/OrderAwareHashMap.scala

```diff
@@ -0,0 +1,800 @@
+    val cap = if(k > 0) ((k + 1).toDouble / defaultLoadFactor).toInt else defaultInitialCapacity
+    new OrderAwareHashMap[K, V](cap, defaultLoadFactor) ++= it
+  }
+
+  def empty[K : Ordering, V]: OrderAwareHashMap[K, V] = new OrderAwareHashMap[K, V]
+
+  def newBuilder[K : Ordering, V]: Builder[(K, V), OrderAwareHashMap[K, V]] = newBuilder(defaultInitialCapacity, defaultLoadFactor)
+
+  def newBuilder[K : Ordering, V](initialCapacity: Int, loadFactor: Double): Builder[(K, V), OrderAwareHashMap[K, V]] =
+    new GrowableBuilder[(K, V), OrderAwareHashMap[K, V]](new OrderAwareHashMap[K, V](initialCapacity, loadFactor)) {
+      override def sizeHint(size: Int) = elems.sizeHint(size)
+    }
+
+  /** The default load factor for the hash table */
+  final def defaultLoadFactor: Double = 0.75
+
+  /** The default initial capacity for the hash table */
+  final def defaultInitialCapacity: Int = 16
+
+  @SerialVersionUID(3L)
+  private final class DeserializationFactory[K, V](val tableLength: Int, val loadFactor: Double, val ordering: Ordering[K]) extends Factory[(K, V), OrderAwareHashMap[K, V]] with Serializable {
+    def fromSpecific(it: IterableOnce[(K, V)]): OrderAwareHashMap[K, V] = new OrderAwareHashMap[K, V](tableLength, loadFactor)(ordering) ++= it
+    def newBuilder: Builder[(K, V), OrderAwareHashMap[K, V]] = OrderAwareHashMap.newBuilder(tableLength, loadFactor)(ordering)
+  }
+
+  @`inline` private def compare[K, V](key: K, hash: Int, node: LLNode[K, V])(implicit ord: Ordering[K]): Int = {
+    val i = hash - node.hash
+    if(i != 0) i else ord.compare(key, node.key)
+  }
+
+  @`inline` private def compare[K, V](key: K, hash: Int, node: RBNode[K, V])(implicit ord: Ordering[K]): Int = {
+    val i = hash - node.hash
+    if(i != 0) i else ord.compare(key, node.key)
+  }
+
+  private final val treeifyThreshold = 8
+
+  // Superclass for RBNode and LLNode to help the JIT with optimizing instance checks, but no shared common fields.
+  // Keeping calls monomorphic where possible and dispatching manually where needed is faster.
+  sealed abstract class Node
+
+  /////////////////////////// Red-Black Tree Node
+
+  final class RBNode[K, V](var key: K, var hash: Int, var value: V, var red: Boolean, var left: RBNode[K, V], var right: RBNode[K, V], var parent: RBNode[K, V]) extends Node {
+    override def toString: String = "RBNode(" + key + ", " + hash + ", " + value + ", " + red + ", " + left + ", " + right + ")"
+
+    @tailrec def getNode(k: K, h: Int)(implicit ord: Ordering[K]): RBNode[K, V] = {
+      val cmp = compare(k, h, this)
+      if (cmp < 0) {
+        if(left ne null) left.getNode(k, h) else null
+      } else if (cmp > 0) {
+        if(right ne null) right.getNode(k, h) else null
+      } else this
+    }
+
+    def foreach[U](f: ((K, V)) => U): Unit = {
+      if(left ne null) left.foreach(f)
+      f((key, value))
+      if(right ne null) right.foreach(f)
+    }
+
+    def foreachNode[U](f: RBNode[K, V] => U): Unit = {
```

Comment:
probably better to write this in terms of successor (for stack safety (if this is at all a concern)). successor could potentially be inlined too.
```
var node = minNodeNonNull(this)
while(node != null) {
  f(node)
  node = successor(node)
}
```

Reply:
Both standalone red-black tree implementations do it the same way. AFAIR using `successor` was considerably slower. Stack safety is not a big concern because red-black trees are always balanced with an upper bound for the height of 2*log2(size+1). With the size limited to Int.MaxValue that should give it a maximum call stack depth of 64 (at which point the amount of heap memory required is probably a much bigger issue).

Reply:
Thanks for the information. I wasn't aware of these properties 😁